[distributions] KL-Divergence for Multivariate Normal

vishwakftw opened the pull request:
This PR implements the KL-Divergence for Multivariate Normal and tests for the same.

Math reviewed and approved by @fritzo [here](https://github.com/probtorch/pytorch/pull/144).

vishwakftw committed:
Add KL-D initial implementation for MVN

vishwakftw committed:
Minor fixes in KL-divergence expression

vishwakftw committed:
Merge branch 'master' into kl-mvn

vishwakftw committed:
Add initial tests

vishwakftw committed:
Complete KL-D tests for Multivariate normal

vishwakftw committed:
Address review comments
1. Add source for the KL-divergence expression

vishwakftw committed:
Fix lint

vishwakftw committed:
Add batched MVN test, and fix KL-D expression error

vishwakftw committed:
Fix lint

vishwakftw committed:
Optimization in result computation

vishwakftw committed:
Random seed change

fritzo commented:
> @vishwakftw  The tests are failing in the CI. However, the tests are not failing locally.

1. Are you sure the tests are running locally?
2. Try merging master into your branch?
3. Try examining the build log and exactly reproducing the environment it builds (maybe there it is using a different lapack or sth).

vishwakftw committed:
Merge branch 'master' into kl-mvn

vishwakftw committed:
Remove usage of q.precision_matrix in favour of _batch_inverse

vishwakftw commented on torch/distributions/kl.py:
```diff
@@ -278,6 +307,21 @@ def _kl_laplace_laplace(p, q):
+
+    term1 = _batch_diag(q.scale_tril).log().sum(-1) - _batch_diag(p.scale_tril).log().sum(-1)
+
+    term2 = _batch_trace(_batch_mm(_batch_inverse(q.covariance_matrix), p.covariance_matrix))
```
I am using this is instead of `q.precision_matrix`. Something seems to have been broken between current HEAD and https://github.com/pytorch/pytorch/commit/c2703aa1412fa07f2dc909abda635770a5c62792 .

fritzo replied:
@fehiepsi do you think your #6128 may be related?

fehiepsi replied:
In my opinion, it is better to use `q.scale_tril` to calculate things related to the inverse of a positive definite matrix.

In this case, we can simplify the computation by using properties of trace:
$ tr( inv(LLt) @ SSt) = tr( inv(Lt) inv(L) S St) = tr( inv(L) S St inv(Lt)) $
So this trace term can be calculated from inv(L) S by squaring it and take sum.

vishwakftw committed:
Fix lint and restore seed to 0

fehiepsi commented on torch/distributions/kl.py:
```diff
@@ -278,6 +308,21 @@ def _kl_laplace_laplace(p, q):
+    term2 = _batch_trace(_batch_mm(_batch_inverse(q.covariance_matrix), p.covariance_matrix))
+
+    term3 = _batch_mahalanobis(q.scale_tril, (q.loc - p.loc))
+    return term1 + 0.5 * (term2 + term3 - p.event_shape[0])
```
I guess here we should have `0.5 * term1` ?

vishwakftw replied:
I thought the trace would involve fewer computations than the sum. Regarding `0.5 * term1`, @fritzo and I have discussed it in the `probtorch/pytorch` https://github.com/probtorch/pytorch/pull/144.

fehiepsi replied:
Thank you @vishwakftw ! Got it ^^

vishwakftw commented:
@fritzo @fehiepsi There seems to be a problem in the `precision_matrix`. Sample code below:
```
>>> import torch as t
>>> a = t.randn(4, 4)
>>> a = t.distributions.transform_to(t.distributions.constraints.lower_cholesky)(a)
>>> m = t.distributions.MultivariateNormal(loc=t.randn(4), scale_tril=a)
>>> m.covariance_matrix

 1.9258  0.3293 -0.0941 -0.6035
 0.3293  0.2726  0.0188  0.2138
-0.0941  0.0188  0.1522  0.0935
-0.6035  0.2138  0.0935  0.6917
[torch.FloatTensor of size (4,4)]

>>> m.precision_matrix

 2.1758 -0.1910 -0.0405 -0.0837
-0.1910  0.6863  0.0517  0.1311
-0.0405  0.0517  0.1432  0.0066
-0.0837  0.1311  0.0066  0.0370
[torch.FloatTensor of size (4,4)]

>>> m.covariance_matrix.mm(m.precision_matrix)

 4.1817 -0.2260 -0.0785 -0.1410
 0.6457  0.1532  0.0049  0.0162
-0.2223  0.0510  0.0272  0.0148
-1.4156  0.3575  0.0535  0.1048
[torch.FloatTensor of size (4,4)]
```

fritzo commented:
@fehiepsi, @vishwakftw's implementation worked before he rebased onto #6128, so it's looking like there is a bug in `.precision_matrix` introduced by #6128. Maybe we can fix the bug in this PR, and add @vishwakftw's example above as a unit test. (sorry if I missed a math error when reviewing #6128).

vishwakftw commented:
I have found the bug in #6128 and fixed it.

vishwakftw committed:
Fix bug in precision_matrix computation, and revert to using q.precision_matrix

fritzo commented:
@vishwakftw Nice work, I stared at that code for 10 minutes and couldn't see an error :blush: Can you add your example as a unit test?

vishwakftw commented:
Sure, I will add a test for verifying inverses and Cholesky decompositions.

fehiepsi commented:
LGTM, thanks for fixing the bug @vishwakftw ! Just a note that
```
term2 = _batch_mm(_batch_inverse(q.scale_tril), p.scale_tril).pow(2).sum(-1)
```
And I think that you can move `_batch_mm` to the implementation of MultiraviteNormal distribution to simplify precision_matrix computation. But this is not necessary. ^^

fritzo reviewed (approved):
LGTM

vishwakftw committed:
Address comments
1. Move _batch_mm to multivariate_normal.py
2. Add MultivariateNormal properties tests

fritzo commented:
@fehiepsi I'm inclined to implement something simple and correct before getting fancy. What you suggest seems incorrect to me:
```py
term2 = _batch_mm(_batch_inverse(q.scale_tril), p.scale_tril).pow(2).sum(-1)  # correct?
```
This is doing an element-wise square, then summing over the last axis. Can you explain why that is equivalent to a matrix-wise `torch.bmm(m, m.transpose(-1, -2)).diag().sum()`?

vishwakftw commented:
@fritzo I think I have figured a work-around to use only `scale_tril`. I am testing it locally.

fehiepsi commented:
@fritzo Assume `m = [x1, x2, ..., xn]` where `x1` is a row vector. Then `(m @ m.t()).diag() = [x1.x1t, x2.x2t, ... xn.xnt]`. I guess you concerned about the batch version of this formula? I will test it in that case.

vishwakftw committed:
Add efficient computation of trace(XX^{T})

fritzo commented on torch/distributions/kl.py:
```diff
@@ -128,13 +128,13 @@ def _x_log_x(tensor):
     mat_size = bmat.size(-1)
-    values = torch.stack([torch.trace(M) for M in bmat.reshape((-1, mat_size, mat_size))])
-    return values.view(bmat.shape[:-2])
+    flat_trace = bmat.reshape(-1, mat_size * mat_size).pow(2).sum(-1)
```
:+1: this looks better, as equivalent to `.pow(2).sum(-1).sum(-1)`

fritzo replied:
@fehiepsi thanks for pointing out this simplification. I had to work out the math, and see you were almost right, just needed one more `.sum(-1)` I believe.

apaszke commented on torch/distributions/multivariate_normal.py:
```diff
@@ -64,6 +64,17 @@ def _batch_inverse(bmat):
     return flat_inv_bmat.view(bmat.shape)
 
 
+def _batch_mm(bmat1, bmat2):
```
Can’t `torch.matmul` handle that?

fehiepsi replied:
`torch.matmul` is slower in my opinion.

vishwakftw replied:
@apaszke I have fixed this in the latest commit.

apaszke commented on test/test_distributions.py:
```diff
@@ -2623,6 +2631,45 @@ def test_kl_monte_carlo(self):
+    def test_kl_multivariate_normal(self):
+        set_rng_seed(0)  # see Note [Randomized statistical tests]
+        n = 5  # Number of tests for multivariate_normal
+        print('Testing KL(MultivariateNormal, MultivariateNormal) using Monte Carlo')
```
Please don’t print anything in the tests

vishwakftw committed:
Address review comments
1. Remove print statements in test_distributions
2. Use torch.matmul instead of _batch_mm
